### PR TITLE
Array constructor fix

### PR DESCRIPTION
**Pull request by KulikovNikita**

### CODE REVIEW

**Review by rlnx — APPROVED** [review]
Approved assuming the corresponding test will be added as part of https://github.com/oneapi-src/oneDAL/pull/1518

File: cpp/oneapi/dal/test/array.cpp [hunk 1]
```diff
@@ -0,0 +1,379 @@
+/*******************************************************************************
```
Comment (michael-smirnov) [inline cpp/oneapi/dal/test/array.cpp:1]: What is the need to change tests here if you planned to do this work in #1518?

**Comment by KulikovNikita** [comment]
Almost green CI http://intel-ci.intel.com/eb8b4c4d-3201-f133-a0e0-8cdcd4b723a1

**Comment by KulikovNikita** [comment]
Green CI: http://intel-ci.intel.com/eb8ba238-e29c-f142-92ab-8cdcd4b723a1

File: cpp/oneapi/dal/backend/primitives/test/ndarray.cpp [hunk 2]
```diff
@@ -489,6 +506,47 @@ TEST_M(ndarray_test, "can assign ndarray", "[ndarray]") {
     sycl::free(x_ptr, queue);
 }
 
+TEST_M(ndarray_test, "can be flattened with host usm", "[ndarray]") {
+    DECLARE_TEST_POLICY(policy);
+    auto& queue = policy.get_queue();
+
+    auto [x, event] = ndarray<float, 2>::ones(queue, { 7, 5 });
```
Comment (rlnx) [inline cpp/oneapi/dal/backend/primitives/test/ndarray.cpp:513]: The test name contradicts with the test itself. By default, `ndarray` uses shared USM for allocation.

File: cpp/oneapi/dal/backend/primitives/test/ndarray.cpp [hunk 3]
```diff
@@ -489,6 +506,47 @@ TEST_M(ndarray_test, "can assign ndarray", "[ndarray]") {
...
+
+    queue
+        .submit([&](sycl::handler& h) {
+            h.memcpy(res_ptr, raw_ptr, (n * m));
+        })
+        .wait_and_throw();
+
+    sycl::free(res_ptr, queue);
+}
+
```
Comment (rlnx) [inline cpp/oneapi/dal/backend/primitives/test/ndarray.cpp:549]: Where is device USM in this test?

File: cpp/oneapi/dal/backend/primitives/test/ndarray.cpp (hunk 2 above)
Comment (KulikovNikita) [reply cpp/oneapi/dal/backend/primitives/test/ndarray.cpp]: Fixed

File: cpp/oneapi/dal/backend/primitives/test/ndarray.cpp (hunk 3 above)
Comment (KulikovNikita) [reply cpp/oneapi/dal/backend/primitives/test/ndarray.cpp]: Fixed

File: cpp/oneapi/dal/backend/primitives/test/ndarray.cpp [hunk 4]
```diff
@@ -489,6 +506,47 @@ TEST_M(ndarray_test, "can assign ndarray", "[ndarray]") {
...
+    float* res_ptr = sycl::malloc_shared<float>(n * m, queue);
+
+    queue
+        .submit([&](sycl::handler& h) {
+            h.memcpy(res_ptr, raw_ptr, (n * m));
+        })
+        .wait_and_throw();
+
+    sycl::free(res_ptr, queue);
```
Comment (rlnx) [inline cpp/oneapi/dal/backend/primitives/test/ndarray.cpp:547]: I suggest to use our wrappers instead raw memory allocation.

```c++
REQUIRE(raw_arr.get_count() == x.get_count());
const auto raw_arr_host = te::matrix<float>::wrap(queue, raw_arr).to_host();
const float*raw_arr_host_ptr = raw_arr_host.get_data();
for (std::int64_t i = 0; i < x.get_count(); ++i) {
    REQUIRE(raw_arr_host_ptr[i] == 1.0f);
}
```

File: cpp/oneapi/dal/backend/primitives/test/ndarray.cpp [hunk 5]
```diff
@@ -416,6 +416,23 @@ TEST("can cast ndarray to ndview", "[ndarray]") {
     REQUIRE(x_view.get_strides() == x.get_strides());
 }
 
+TEST_M(ndarray_test, "can be flattened host", "[ndarray]") {
```
Comment (rlnx) [inline cpp/oneapi/dal/backend/primitives/test/ndarray.cpp:419]:
```suggestion
TEST_M(ndarray_test, "can be flattened on host", "[ndarray]") {
```

File: cpp/oneapi/dal/backend/primitives/test/ndarray.cpp [hunk 6]
```diff
@@ -416,6 +416,23 @@ TEST("can cast ndarray to ndview", "[ndarray]") {
...
+    auto x = ndarray<float, 2>::wrap(data_vector.data(), { m, n });
+
+    auto raw_arr = x.flatten();
+
+    REQUIRE(raw_arr.get_count() == (m * n));
+
+    for (std::int64_t i = 0; i < (m * n); ++i) {
+        REQUIRE(raw_arr[i] == 1.0f);
+    }
+}
```
Comment (rlnx) [inline cpp/oneapi/dal/backend/primitives/test/ndarray.cpp:434]: Comment regarding separation on code blocks in unit tests. Now it's difficult to follow test flow. Here is some good practice.
https://softwareengineering.stackexchange.com/a/15934

```c++
TEST("...") {
    // Some lines 
    // of setup code
    // go here.

    // The action being tested happens after a blank line.

    // An assertion follows another blank line.
}
```

This is applicable to all unit tests.

File: cpp/oneapi/dal/backend/primitives/test/ndarray.cpp (hunk 4 above)
Comment (KulikovNikita) [reply cpp/oneapi/dal/backend/primitives/test/ndarray.cpp]: I suppose that it this test should be simple and not require additional entities.

File: cpp/oneapi/dal/backend/primitives/test/ndarray.cpp [hunk 7]
```diff
@@ -489,6 +503,42 @@ TEST_M(ndarray_test, "can assign ndarray", "[ndarray]") {
...
+
+TEST_M(ndarray_test, "can be flattened with device usm", "[ndarray]") {
+    DECLARE_TEST_POLICY(policy);
+    auto& queue = policy.get_queue();
+    constexpr std::int64_t m = 7;
+    constexpr std::int64_t n = 5;
+    auto [x, event] = ndarray<float, 2>::ones(queue, { m, n }, sycl::usm::alloc::device);
+    event.wait_and_throw();
+
+    const auto raw_arr = x.flatten();
```
Comment (michael-smirnov) [inline cpp/oneapi/dal/backend/primitives/test/ndarray.cpp:527]: Why host `flatten()` does not return host data as other methods in data management layer?

File: cpp/oneapi/dal/backend/primitives/test/ndarray.cpp [hunk 8]
```diff
@@ -489,6 +503,42 @@ TEST_M(ndarray_test, "can assign ndarray", "[ndarray]") {
...
+    event.wait_and_throw();
+
+    auto raw_arr = x.flatten();
+
+    for (std::int64_t i = 0; i < x.get_count(); ++i) {
+        REQUIRE(raw_arr[i] == 1.0f);
+    }
+}
+
+TEST_M(ndarray_test, "can be flattened with device usm", "[ndarray]") {
```
Comment (michael-smirnov) [inline cpp/oneapi/dal/backend/primitives/test/ndarray.cpp:519]: Why we have different code to test host USM and device USM? We can have single code to test all three types of USM memory

File: cpp/oneapi/dal/test/array.cpp [hunk 9]
```diff
@@ -327,6 +327,45 @@ TEST("can wrap const data with queue, offset and deleter") {
...
+    constexpr std::int64_t count = 3;
+    auto* data = sycl::malloc_shared<float>(count, q);
+    for (std::int64_t i = 0; i < count; ++i) {
+        data[i] = float(i);
+    }
+    std::shared_ptr<float> sdata(data, [&q](auto data) {
+        sycl::free(data, q);
+    });
+
+    auto darr = array<float>{ sdata, count };
```
Comment (michael-smirnov) [inline cpp/oneapi/dal/test/array.cpp:343]: That is not right way to create an array from USM memory. Use ctor with queue parameter instead.

File: cpp/oneapi/dal/test/array.cpp [hunk 10]
```diff
@@ -327,6 +327,45 @@ TEST("can wrap const data with queue, offset and deleter") {
     REQUIRE(arr.get_data() == cdata);
     REQUIRE(!arr.has_mutable_data());
 }
+
+TEST("can be created from shared_ptr") {
```
Comment (michael-smirnov) [inline cpp/oneapi/dal/test/array.cpp:331]:
```suggestion
TEST("can be created from shared_ptr with shared USM data") {
```

**Review by michael-smirnov — CHANGES_REQUESTED** [review]

File: cpp/oneapi/dal/backend/primitives/test/ndarray.cpp [hunk 11]
```diff
@@ -489,6 +507,41 @@ TEST_M(ndarray_test, "can assign ndarray", "[ndarray]") {
...
+    DECLARE_TEST_POLICY(policy);
+    auto& queue = policy.get_queue();
+    constexpr std::int64_t m = 7;
+    constexpr std::int64_t n = 5;
+    INFO("CHECK");
```
Comment (rlnx) [inline cpp/oneapi/dal/backend/primitives/test/ndarray.cpp:528]: This INFO is useless, remove it or figure out more descriptive message

File: cpp/oneapi/dal/backend/primitives/test/ndarray.cpp [hunk 12]
```diff
@@ -489,6 +507,41 @@ TEST_M(ndarray_test, "can assign ndarray", "[ndarray]") {
...
+    auto [x, event] = ndarray<float, 2>::ones(queue, { m, n }, sycl::usm::alloc::device);
+    event.wait_and_throw();
+
+    const auto raw_arr = x.flatten(queue);
+
+    REQUIRE(raw_arr.get_count() == x.get_count());
+    INFO("CHECK");
```
Comment (rlnx) [inline cpp/oneapi/dal/backend/primitives/test/ndarray.cpp:535]: This INFO is useless, remove it or figure out more descriptive message

File: cpp/oneapi/dal/backend/primitives/test/ndarray.cpp (hunk 4 above)
Comment (KulikovNikita) [reply cpp/oneapi/dal/backend/primitives/test/ndarray.cpp]: Fixed

File: cpp/oneapi/dal/backend/primitives/test/ndarray.cpp (hunk 7 above)
Comment (KulikovNikita) [reply cpp/oneapi/dal/backend/primitives/test/ndarray.cpp]: Cause it has only the pointer.

File: cpp/oneapi/dal/backend/primitives/test/ndarray.cpp (hunk 11 above)
Comment (KulikovNikita) [reply cpp/oneapi/dal/backend/primitives/test/ndarray.cpp]: Fixed

File: cpp/oneapi/dal/backend/primitives/test/ndarray.cpp (hunk 12 above)
Comment (KulikovNikita) [reply cpp/oneapi/dal/backend/primitives/test/ndarray.cpp]: Fixed

File: cpp/oneapi/dal/test/array.cpp [hunk 13]
```diff
@@ -327,6 +360,7 @@ TEST("can wrap const data with queue, offset and deleter") {
     REQUIRE(arr.get_data() == cdata);
     REQUIRE(!arr.has_mutable_data());
 }
+
```
Comment (rlnx) [inline cpp/oneapi/dal/test/array.cpp:363]:
```suggestion
```

**Review by rlnx — APPROVED** [review]

**Review by michael-smirnov — APPROVED** [review]
Assuming ndarray/flatten issues will be fixed in next PRs

**Comment by KulikovNikita** [comment]
Green CI: http://intel-ci.intel.com/eb8d8df4-bdda-f190-96e6-8cdcd4b723a1

**Pull request merged**